fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/SlotProvider.java

```diff
@@ -0,0 +1,129 @@
+import com.starrocks.common.util.DebugUtil;
+import com.starrocks.rpc.FrontendServiceProxy;
+import com.starrocks.thrift.TReleaseSlotRequest;
+import com.starrocks.thrift.TReleaseSlotResponse;
+import com.starrocks.thrift.TRequireSlotRequest;
+import com.starrocks.thrift.TStatusCode;
+import com.starrocks.thrift.TUniqueId;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class SlotProvider {
+    private static final Logger LOG = LogManager.getLogger(SlotProvider.class);
+
+    private final ConcurrentMap<TUniqueId, PendingSlotRequest> pendingSlots = new ConcurrentHashMap<>();
+
+    public CompletableFuture<Slot> requireSlot(Slot slot) {
+        PendingSlotRequest slotRequest = new PendingSlotRequest(slot);
+        slotRequest.onRequire();
+
+        pendingSlots.put(slot.getSlotId(), slotRequest);
+
+        try {
+            requireSlotFromSlotManager(slot);
+        } catch (Exception e) {
+            LOG.warn("[Slot] failed to require slot [slot={}]", slot, e);
+            pendingSlots.remove(slot.getSlotId());
+            slotRequest.onFailed(e);
+        }
+
+        return slotRequest.getSlotFuture();
+    }
+
+    public Status finishSlotRequirement(TUniqueId slotId, Status status) {
+        PendingSlotRequest slotRequest = pendingSlots.remove(slotId);
+        if (slotRequest == null) {
+            LOG.warn("[Slot] finishSlotRequirement receives a response with non-exist slotId [slotId={}] [status={}]",
+                    DebugUtil.printId(slotId), status);
+            return Status.internalError("the slotId does not exist");
+        }
+
+        if (status.ok()) {
+            slotRequest.onFinished();
+        } else {
+            LOG.warn("[Slot] finishSlotRequirement receives a failed response [slot={}] [status={}]", slotRequest, status);
+            slotRequest.onFailed(new UserException(status.getErrorMsg()));
+        }
+
+        return new Status();
+    }
+
+    public void cancelSlotRequirement(Slot slot) {
+        if (slot == null) {
+            return;
+        }
+
+        PendingSlotRequest slotRequest = pendingSlots.remove(slot.getSlotId());
+        if (slotRequest == null) {
+            return;
+        }
+
+        slotRequest.onCancel();
+        releaseSlotToSlotManager(slot);
+    }
+
+    public void releaseSlot(Slot slot) {
+        if (slot == null || slot.getState() != Slot.State.ALLOCATED) {
+            return;
+        }
+
+        slot.onRelease();
+        releaseSlotToSlotManager(slot);
+    }
+
+    private void requireSlotFromSlotManager(Slot slot) throws Exception {
+        TRequireSlotRequest request = new TRequireSlotRequest();
+        request.setSlot(slot.toThrift());
+
+        FrontendServiceProxy.call(slot.getLeaderEndpoint(),
```

Comment:
Would Leader FE also require slots via RPC?
Would Leader FE be busy in serving overwhelming requests of requireSlots?

Reply:
1. Yes, Leader FE also vida RPC to make the control flow consistent.
2. The leader FE only uses one thread to dealing with requests of requireSlots. Each request operation is simple, which cannot take too much time. When a request arrives at the leader, if the current queued request exceed `query_queue_max_queued_queries` (1024) by default, this request will be failed immediately.